boileroom/models/sae/core.py

```
@@ -0,0 +1,392 @@
+
+        pooled_list: list[np.ndarray] = []
+        residue_feature_list: list[np.ndarray] = []
+        num_features = 0
+        with Timer("Forge SAE features") as forge_timer:
+            for item in parsed:
+                dense = np.asarray(self._forge.features(item.sdk_sequence), dtype=np.float32)
+                if dense.ndim != 2:
+                    raise ValueError(f"Forge features must be 2-D (tokens, num_features); got shape {dense.shape}.")
+                # Residue token positions: BOS is index 0, chain-break '|' tokens are
+                # dropped, EOS is the trailing token.
+                keep = [i + 1 for i, token in enumerate(item.sdk_sequence) if token != "|"]
```

Ruff (0.16.1)

[error] Possible hardcoded password assigned to: "token"

(S105)